lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+                  mem = parentView.getViewSource();
+
+                for (auto alias : aliases.resolve(mem)) {
+                  auto op = alias.getDefiningOp();
+                  if (op) {
+                    if (mlir::isa<mlir::memref::GetGlobalOp>(op)) {
+                      gpuGetMemrefGlobalParams.insert({op, {}});
+                      continue;
+                    }
+                    if (op->getDialect() == scfDialect ||
+                        mlir::isa<mlir::ViewLikeOpInterface>(op))
+                      continue;
+                    if (mlir::isa<mlir::memref::AllocOp>(op)) {
+                      gpuBufferAllocs.insert({op, {}});
+                    } else if (mlir::isa<mlir::func::CallOp>(op)) {
+                      // Ignore
+                    } else {
+                      op->emitError("Unhandled memref producer");
+                      return mlir::WalkResult::interrupt();
+                    }
+
+                  } else {
+                    auto block = alias.getParentBlock();
+                    auto blockArgs = block->getArguments();
+                    auto it = llvm::find(blockArgs, alias);
+                    assert(it != blockArgs.end());
+                    auto index = static_cast<unsigned>(it - blockArgs.begin());
+                    gpuBufferParams.insert({index, {}});
+                  }
+                }
+              }
+
+              return mlir::WalkResult::advance();
+            })
+            .wasInterrupted()) {
+      signalPassFailure();
+      return;
+    }
+
+    // Checks the access type of the OP
+    auto getAccessType = [&](mlir::Value memref) {
+      AccessType ret;
+      for (auto mem : aliases.resolve(memref)) {
+        for (auto user : mem.getUsers()) {
+          if (mlir::isa<mlir::func::ReturnOp>(user)) {
+            ret.hostRead = true;
+            ret.hostWrite = true;
+            continue;
+          }
+
+          if (auto copy = mlir::dyn_cast<mlir::memref::CopyOp>(user)) {
+            if (copy.source() == mem)
+              ret.hostRead = true;
+
+            if (copy.target() == mem)
+              ret.hostWrite = true;
+
+            continue;
+          }
+
+          if (auto memInterface =
+                  mlir::dyn_cast<mlir::MemoryEffectOpInterface>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            if (memInterface.hasEffect<mlir::MemoryEffects::Read>())
+              (onDevice ? ret.deviceRead : ret.hostRead) = true;
+
+            if (memInterface.hasEffect<mlir::MemoryEffects::Write>())
+              (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
+
+            continue;
+          }
+          if (mlir::isa<mlir::func::CallOp>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            (onDevice ? ret.deviceRead : ret.hostRead) = true;
+            (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
```

Comment:
What is the use for "deviceRead " and "deviceWrite"? Do the code do any optimization for these tags?